src/hotspot/share/gc/shared/preservedMarks.hpp

```diff
@@ -34,27 +34,30 @@ class WorkerTask;
 class PreservedMarksSet;
 class WorkerThreads;
 
+class OopAndMarkWord {
+ private:
+  oop _o;
+  markWord _m;
+
+ public:
+  OopAndMarkWord(oop obj, markWord m) : _o(obj), _m(m) { }
+
+  oop get_oop() { return _o; }
+  oop* get_oop_pointer() {return &_o; }
+  inline void set_mark() const;
+  void set_oop(oop obj) { _o = obj; }
+};
+
 class PreservedMarks {
 private:
-  class OopAndMarkWord {
-  private:
-    oop _o;
-    markWord _m;
-
-  public:
-    OopAndMarkWord(oop obj, markWord m) : _o(obj), _m(m) { }
-
-    oop get_oop() { return _o; }
-    inline void set_mark() const;
-    void set_oop(oop obj) { _o = obj; }
-  };
   typedef Stack<OopAndMarkWord, mtGC> OopAndMarkWordStack;
 
   OopAndMarkWordStack _stack;
 
   inline bool should_preserve_mark(oop obj, markWord m) const;
 
 public:
+  OopAndMarkWordStack& get_stack() { return _stack; }
```

Comment:
The methods `MarkSweep::adjust_pointer` and `PreservedMarks::adjust_during_full_gc` have different implementations. The method `MarkSweep::adjust_pointer` seems doing more thing. So in `MarkSweep::adjust_marks`, I use `get_stack` to get the stack and invoke `MarkSweep::adjust_pointer` explicitly. If you can confirm `PreservedMarks::adjust_during_full_gc` is good to replace `MarkSweep::adjust_pointer`, I will remove `get_stack` and use `PreservedMarks::adjust_during_full_gc` instead of `MarkSweep::adjust_pointer`.

Reply:
You can replace that code with `PreservedMarks::adjust_during_full_gc` from what I can see; the additional code in `MarkSweep::adjust_pointer` only deals with the generic case if the passed `oop` is null, which can't be the case for oops in the preserved marks set.